core/src/Kokkos_Tuners.hpp

```diff
@@ -0,0 +1,218 @@
+     */
+    std::vector<int64_t> configuration_indices;
+
+    if (policy.auto_vector_length()) {
+      policy.impl_set_vector_length(1);  // TODO: find a heuristic
+    }
+
+    // in all cases, start by pushing the default
+    configuration_indices.push_back(0);  // default option index
+    if (policy.auto_team_size()) {       // case 1 or 3
+      auto default_team_size = policy.team_size_recommended(functor, tag);
+      configurations.push_back(
+          std::make_pair(default_team_size, policy.vector_length()));
+    } else {  // case 2
+      configurations.push_back(
+          std::make_pair(policy.team_size(), policy.vector_length()));
+    }
+
+    auto max_vector_length = PolicyType::vector_length_max();
+    std::vector<int64_t> allowed_vector_lengths;
+
+    if (policy.auto_vector_length()) {  // case 1 or 2
+      for (int vector_length = max_vector_length; vector_length >= 1;
+           vector_length /= 2) {
+        policy.impl_set_vector_length(vector_length);
+        /**
+         * Figuring out whether a vector length is valid depends
+         * on whether we're in case 1 (tune everything) or 2 (just tune vector
+         * length)
+         *
+         * If we're tuning everything, all legal vector lengths are valid.
+         * If we're just tuning vector length, we need to check that if we
+         * set this vector length, the team size provided will be valid.
+         *
+         * These are the left and right hand sides of the "or" in this
+         * conditional, respectively.
+         */
+        auto max_team_size = policy.team_size_max(functor, tag);
+        if ((policy.auto_team_size()) ||
+            (policy.team_size() <= max_team_size)) {
+          allowed_vector_lengths.push_back(vector_length);
+        }
+      }
+    } else {  // case 3, there's only one vector length to care about
+      allowed_vector_lengths.push_back(policy.vector_length());
+    }
+    int index = 1;
+
+    for (const auto vector_length : allowed_vector_lengths) {
+      policy.impl_set_vector_length(vector_length);
+      auto max_team_size = policy.team_size_max(functor, tag);
+      if (policy.auto_team_size()) {  // case 1 or 3, try all legal team sizes
+        for (int team_size = max_team_size; team_size >= 1; team_size /= 2) {
+          configuration_indices.push_back(index++);
+          configurations.push_back(std::make_pair(team_size, vector_length));
+        }
+      } else {  // case 2, just try the provided team size
+        configuration_indices.push_back(index++);
+        configurations.push_back(
+            std::make_pair(policy.team_size(), vector_length));
+      }
+    }
+
+    declare_output_type(name, configuration_indices);
+    policy.impl_set_vector_length(initial_vector_length);
+  }
+
+  template <typename... Properties>
+  void tune(Kokkos::TeamPolicy<Properties...>& policy,
+            const size_t context_id) {
+    Kokkos::Tools::Experimental::VariableValue value_index{type_id, 0};
```

Comment:
We should add comments here (assuming I'm understanding this correctly) that 0 is the default Kokkos heuristic. And that request_values does nothing unless something is actually loaded.

Reply:
Fixed